social_engagement/handlers.py

```diff
@@ -98,20 +115,85 @@ def comment_created_signal_handler(sender, **kwargs):  # pylint: disable=unused-
 
             # update the engagement score of the thread creator
             # as well
-            task_update_user_engagement_score.delay(unicode(course_id), thread_user_id)
+            _increment(thread_user_id, course_id, 'num_comments_generated')
 
 
 @receiver(thread_followed)
+def thread_follow_signal_handler(sender, **kwargs):  # pylint: disable=unused-argument
+    """
+    Updates user social engagement score
+    """
+    thread = kwargs['post']
+    course_id = getattr(thread, 'course_id', None)
+    user_id = getattr(thread, 'user_id', None)
+    action_user = kwargs['user']  # user who followed or un-followed thread
+
+    if user_id and action_user.id != int(user_id):
+        _increment(user_id, course_id, 'num_thread_followers')
+
+
 @receiver(thread_unfollowed)
-def thread_follow_unfollow_signal_handler(sender, **kwargs):  # pylint: disable=unused-argument
+def thread_unfollow_signal_handler(sender, **kwargs):  # pylint: disable=unused-argument
     """
     Updates user social engagement score
     """
     thread = kwargs['post']
     course_id = getattr(thread, 'course_id', None)
     user_id = getattr(thread, 'user_id', None)
     action_user = kwargs['user']  # user who followed or un-followed thread
-    social_engagement_enabled = settings.FEATURES.get('ENABLE_SOCIAL_ENGAGEMENT', False)
 
-    if social_engagement_enabled and course_id and user_id and action_user.id != user_id:
-        task_update_user_engagement_score.delay(unicode(course_id), user_id)
+    if user_id and action_user.id != int(user_id):
+        # task_update_user_engagement_score.delay(unicode(course_id), user_id)
+        _decrement(user_id, course_id, 'num_thread_followers')
+
+
+def _increment(*args, **kwargs):
+    """
+    A facade for handling incrementation.
+    """
+    _change_handler(*args, **kwargs)
+
+
+def _decrement(*args, **kwargs):
+    """
+    A facade for handling decrementation.
+    """
+    _change_handler(*args, increment=False, **kwargs)
+
+
+def _change_handler(user_id, course_id, param, increment=True, items=1):
+    """
+    Save changes and calculate score.
+    """
+    factor = items if increment else -items
+    social_metric_points = getattr(
+        settings,
+        'SOCIAL_METRIC_POINTS',
+        {
+            'num_threads': 10,
+            'num_comments': 15,
+            'num_replies': 15,
+            'num_upvotes': 25,
+            'num_thread_followers': 5,
+            'num_comments_generated': 15,
+        }
```

Comment:
This seems to be used elsewhere, too. If these defaults changed, they'd have to be changed everywhere. Can you factor the default dictionary out to a common constant that can be used everywhere?

Reply:
Also this seems to be the default in Apros and comes from there. I haven't looked into the code carefully enough to understand why it's being used in both places, but that's a bit weird to me.

Reply:
@UmanShahzad, in one place it was used by management command and in another one it's used after sending signals. I'm moving it to an external function.